src/Symfony/Component/Translation/Tests/TranslatorCacheTest.php

```diff
@@ -51,126 +52,133 @@ protected function deleteTmpDir()
-        $this->assertEquals('bar (EN)', $translator->trans('bar'));
-        $this->assertEquals('foobar (ES)', $translator->trans('foobar'));
-        $this->assertEquals('choice 0 (EN)', $translator->transChoice('choice', 0));
-        $this->assertEquals('no translation', $translator->trans('no translation'));
-        $this->assertEquals('foobarfoo (PT-PT)', $translator->trans('foobarfoo'));
-        $this->assertEquals('other choice 1 (PT-BR)', $translator->transChoice('other choice', 1));
-        $this->assertEquals('foobarbaz (fr.UTF-8)', $translator->trans('foobarbaz'));
-        $this->assertEquals('foobarbax (sr@latin)', $translator->trans('foobarbax'));
+        $locale = 'whatever';
+        $format = 'what-you-like';
+        $msgid = 'test';
+
+        // Prime the cache
+        $translator = new Translator($locale, null, $this->tmpDir, $debug);
+        $translator->addLoader($format, new ArrayLoader());
+        $translator->addResource($format, array($msgid => 'OK'), $locale);
+        $translator->trans($msgid);
+
+        // Try again and see we get a valid result whilst no loader can be used
+        $translator = new Translator($locale, null, $this->tmpDir, $debug);
+        $translator->addLoader($format, $this->createFailingLoader());
+        $translator->addResource($format, array($msgid => 'OK'), $locale);
+        $this->assertEquals('OK', $translator->trans($msgid), '-> caching does not work in '.($debug ? 'debug' : 'production'));
     }
 
-    public function testRefreshCacheWhenResourcesChange()
+    public function testCatalogueIsReloadedWhenResourcesAreNoLongerFresh()
     {
-        // prime the cache
-        $loader = $this->getMock('Symfony\Component\Translation\Loader\LoaderInterface');
-        $loader
-            ->method('load')
-            ->will($this->returnValue($this->getCatalogue('fr', array(
-                'foo' => 'foo A',
-            ))))
-        ;
+        /*
+         * The testThatACacheIsUsed() test showed that we don't need the loader as long as the cache
+         * is fresh.
+         *
+         * Now we add a Resource that is never fresh and make sure that the
+         * cache is discarded (the loader is called twice).
+         *
+         * We need to run this for debug=true only because in production the cache
+         * will never be revalidated.
+         */
 
-        $translator = new Translator('fr', new MessageSelector(), $this->tmpDir, true);
-        $translator->setLocale('fr');
-        $translator->addLoader('loader', $loader);
-        $translator->addResource('loader', 'foo', 'fr');
+        $locale = 'whatever';
+        $format = 'what-you-like';
+        $msgid = 'test';
 
-        $this->assertEquals('foo A', $translator->trans('foo'));
+        $catalogue = new MessageCatalogue($locale, array());
+        $catalogue->addResource(new StaleResource()); // better use a helper class than a mock, because it gets serialized in the cache and re-loaded
```

Comment:
we can use mock here the only thing that doesn't work as expected is  `expects` method.

```php
$resource = $this->getMock('Symfony\Component\Config\Resource\ResourceInterface');
$resource->method('isFresh')->will($this->returnValue(false));
catalogue->addResource($resource);
```